Statement averaged scores

scottagt opened the pull request:
CJ, could you have a look and let me know your thoughts? These are the changes I mentioned last week for doing average bias of a statement. It uses sent tokenize from nltk to get sentences and then uses multiprocessing to parallelize the bias scoring of each sentence before averaging together.

Also, I mentioned it earlier in email I think but, I had to import different pattern.en dependencies to get it to work during development but suspect it should work the same if the import sources for pattern.en are not changed assuming same functions are available.

scottagt committed:
whole statement average bias method
updating imports for latest pattern.en

scottagt committed:
adding multiprocessing support

jpfairbanks commented on bs_detector/bias.py:
```diff
@@ -19,6 +21,12 @@ def get_list_from_file(file_name):
     return l
 
 
+def get_text_from_article_file(article_file):
```
I think this argument should be `article_file_path` since you call `open` on it. `article_file` to me sounds like I should have already opened the file.

scottagt replied:
Agree

jpfairbanks commented on bs_detector/bias.py:
```diff
@@ -475,27 +483,59 @@ def print_raw_data_for_features(list_of_sentences):
-#print_raw_data_for_features(get_list_from_file('input_text_original'))
+
+
+# print_raw_data_for_features(get_list_from_file('input_text_original'))
```
this dead code can be deleted

scottagt replied:
Ask @cjhutto about it. My IDE fixed some spacing with the comment sign is all that change was.

cjhutto replied:
it was already deleted on the most recent update (where I added html output and TSV output formats)

scottagt replied:
Ah ok, I may have pulled right before your latest changes. Will update :)

jpfairbanks replied:
Ok then it will get deleted when merge conflicts are handled.

jpfairbanks commented on bs_detector/bias.py:
```diff
@@ -475,27 +483,59 @@ def print_raw_data_for_features(list_of_sentences):
+    if isinstance(sentence_text, unicode):
+        features = extract_bias_features(sentence_text)
+    else:
+        features = extract_bias_features(unicode(sentence_text, errors='ignore'))
```
Should this unicode handling be built into the `extract_bias_features` function? It seems like it will be repeated every time you call `extract_bias_features`?

scottagt replied:
Its related to a change with:
sentences = nltk.sent_tokenize(statement_text.decode("ascii", "ignore"))

where the text in the file is an unknown encoding and the sentences are being converted to ascii for sentence tokenization. So then afterwards, the sentence need to be converted back to unicode (with errors=ignore).

It looks like there is a unicode call within extract_bias_feature but not with 'errors=ignore'. If @cjhutto agrees I'll update that to use errors=ignore or let's discuss other thoughts or assumptions about inputs. One concern for the case of multi-sentence statements is to give the best input to the nltk sentence tokenization method.

cjhutto replied:
The unicode handling was moved to be built into the `extract_bias_features` function (for a similar reason mentioned -- it is needed for the nltk tokenization)...

jpfairbanks commented on bs_detector/bias.py:
```diff
@@ -475,27 +483,59 @@ def print_raw_data_for_features(list_of_sentences):
+
+    with poolcontext(processes=n_jobs) as pool:
+        bs_scores = pool.map(compute_bias, sentences)
+        avg_bias = sum(bs_scores)
```
Using multiprocessing here will only be faster if the number of sentences is `yoooge`. Have you done any profiling to see if it is necessary on reasonably sized documents?

I would rather make the code more efficient that throw processors at the problem if we can.

scottagt replied:
No, no profiling yet but thanks for suggesting, agree. The rationale was that the user could decide and edit the n_jobs parameter to suit their needs. Better to make the option available. I think we can have this and also continue to make more efficient code so no need to choose one or the other. But so are you saying that you don't think an article with 200 sentences wouldn't see a performance increase over serial? I tend to think there would be some performance increase on a multi-core system. And what do you consider yooge? 1K 1M 10M?

jpfairbanks replied:
Yeah I think 1K is the minimum at which I would parallelize. You have to copy all the data into the subprocesses which should take a big chunk of the time in processing them. If you make a process pool with one job does it actually make a subprocess and do all the overhead? Or do they use a dummy process pool that does the work in the main process?

jpfairbanks replied:
Of course I would want to profile and find the hot spots and optimize them, before I started to parallelize within a single document. It is much more efficient to parallelize over the independent documents if that is possible.

scottagt replied:
Ok so, I ran profiling and I do see a performance improvement in terms of total time. See attached. Thoughts? I've also created a new non-MP method in the process, so we'll have that either way now.

MP (7494ms)
<img width="1153" alt="compute_statement_mp" src="https://user-images.githubusercontent.com/25912784/33329519-be3278b8-d42a-11e7-9895-1dfa57b32ad1.png">

No MP (22179ms)
<img width="1296" alt="compute_statement_nomp" src="https://user-images.githubusercontent.com/25912784/33329524-c429b97a-d42a-11e7-92f1-53fde9d2c0db.png">

scottagt replied:
Oh, also, I think the parallelization could benefit at the document and also the within document levels since sentences are independent units of bias scoring and if we agree on the above interpretation then there are gains to be made with MP on statement-level (a document can be thought of in some sense as a big statement)

jpfairbanks replied:
I am pleasantly surprised about that 3x on (i assume 4 cores). I see an easy pick up in accelerating the serial implementation.

jpfairbanks commented on bs_detector/bias.py:
```diff
@@ -475,27 +483,59 @@ def print_raw_data_for_features(list_of_sentences):
+        avg_bias = sum(bs_scores)
+
+    if len(sentences) > 0:
+        avg_bias = round(float(avg_bias) / float(len(sentences)), 4)
```
Why are we rounding to 4 digits? Is that a significant figures thing? Do we not trust the 1e-5 component to be accurate?

scottagt replied:
This is keeping in line with @cjhutto's earlier code where round is also used. How do you feel about those lines too? I'm not tied to doing it, just wanting to keep in line with earlier convention.

jpfairbanks replied:
Yeah it is good to stick with convention @cjhutto why where you rounding? Just for making pretty tables?

cjhutto replied:
The input for bias ratings are integers -- four decimal places is waaaay more than enough precision.
Oh, and yes...Pretty Tables :)

jpfairbanks reviewed (commented):
Hi @scottagt,

Thanks for this pull request, I have added some notes and questions. Can you let me know what you think?

scottagt commented:
Thanks @jpfairbanks for comments :)

scottagt committed:
adding article

scottagt committed:
Merge branch 'master' into statement_averaged_scores

# Conflicts:
#	bs_detector/bias.py

scottagt committed:
adding new non-MP function
updating unicode processing to handle non-unicode and unicode sentences after sentence tokenization

jpfairbanks commented:
@cjhutto @scottagt LGTM? 🚢

scottagt commented:
Hi @cjhutto do you approve too?

jpfairbanks commented:
@cjhutto Lets get a decision on this today. Will merge by COB

jpfairbanks commented:
🍾 🎆
